src/server.zig

```diff
@@ -0,0 +1,293 @@
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+const std = @import("std");
+
+const public = @import("jsruntime");
+const Completion = public.IO.Completion;
+const AcceptError = public.IO.AcceptError;
+const RecvError = public.IO.RecvError;
+const SendError = public.IO.SendError;
+const TimeoutError = public.IO.TimeoutError;
+
+const MsgBuffer = @import("msg.zig").MsgBuffer;
+const Browser = @import("browser/browser.zig").Browser;
+const cdp = @import("cdp/cdp.zig");
+
+const NoError = error{NoError};
+const IOError = AcceptError || RecvError || SendError || TimeoutError;
+const Error = IOError || std.fmt.ParseIntError || cdp.Error || NoError;
+
+// I/O Recv
+// --------
+
+const BufReadSize = 1024; // 1KB
+const MaxStdOutSize = 512; // ensure debug msg are not too long
+
+pub const Cmd = struct {
+
+    // internal fields
+    socket: std.posix.socket_t,
+    buf: []u8, // only for read operations
+    err: ?Error = null,
+
+    msg_buf: *MsgBuffer,
+
+    // CDP
+    state: cdp.State = .{},
+
+    // JS fields
+    browser: *Browser, // TODO: is pointer mandatory here?
+    // try_catch: public.TryCatch, // TODO
+
+    fn cbk(self: *Cmd, completion: *Completion, result: RecvError!usize) void {
+        const size = result catch |err| {
+            self.err = err;
+            return;
+        };
+
+        if (size == 0) {
+            // continue receving incomming messages asynchronously
+            self.loop().io.recv(*Cmd, self, cbk, completion, self.socket, self.buf);
+            return;
+        }
+
+        // input
+        const input = self.buf[0..size];
+        if (std.log.defaultLogEnabled(.debug)) {
+            const content = input[0..@min(MaxStdOutSize, size)];
+            std.debug.print("\ninput size: {d}, content: {s}\n", .{ size, content });
+        }
+
+        // close on exit command
+        if (std.mem.eql(u8, input, "exit")) {
+            self.err = error.NoError;
+            return;
+        }
+
+        // read and execute input
+        self.msg_buf.read(self.alloc(), input, self, Cmd.do) catch unreachable;
+
+        // continue receving incomming messages asynchronously
+        self.loop().io.recv(*Cmd, self, cbk, completion, self.socket, self.buf);
+    }
+
+    // shortcuts
+    inline fn alloc(self: *Cmd) std.mem.Allocator {
+        // TODO: should we return the allocator from the page instead?
+        return self.browser.currentSession().alloc;
```

Comment:
The session initialize the js env with an arena created from the session's allocator.
The page starts and stops the env and free its own arena and the libdom memory.

Since`Cmd` allows navigation through multiple pages, It makes more sense to tie it with the browser's allocator, env and loop.

My real question is: It makes more sense to inject a `*Session` into `Cmd` instead of an `*Browser` itself.
Then you will be able to remove all these `browser.currentSession().XX`

Reply:
We need to keep a reference to the browser to allow create/delete sessions for multiple CDP calls on 1 connection